core/include/cond.h

```diff
@@ -0,0 +1,223 @@
+/*
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @brief       Condition variable for thread synchronization
+ * @ingroup     core
+ * @{
+ *
+ * @file
+ * @brief       RIOT synchronization API
+ *
+ * This file contains a condition variable with Mesa-style semantics.
+ *
+ * Condition variable solve the following problem. Suppose that a thread should
+ * sleep until a certain condition comes true. Condition variable provide a
+ * primitive whereby a thread can go to sleep by calling cond_wait(). Then,
+ * when the condition comes true in a thread or interrupt context, cond_signal()
+ * can be called, to wake up the thread.
+ *
+ * "Mesa-style semantics" means that, when cond_signal() is called, the
+ * sleeping thread becomes runnable, but may not be scheduled immediately. In
+ * contrast, "Hoare-style semantics" means that when cond_signal() is called,
+ * the sleeping thread is awakened and immediately scheduled. The condition
+ * variable in this file implements Mesa-style semantics, as is used by other
+ * standard implementations, such as pthreads.
+ *
+ * To avoid races, condition variables are used with mutexes. When a thread is
+ * put to sleep with cond_wait, it atomically unlocks the provided mutex and
+ * then goes to sleep. When it is awakened with cond_signal, it reacquires the
+ * mutex.
+ *
+ * As a rule of thumb, every condition variable should have a corresponding
+ * mutex, and that mutex should be held whenever performing any operation with
+ * the condition variable. There are exceptions to this rule, where it is
+ * appropriate to call cond_signal or cond_broadcast without the mutex held
+ * (for example, if you know that no thread will call cond_wait concurrently).
+ * It is safe to call cond_signal or cond_broadcast in interrupt context.
+ *
+ * However, the programmer should be aware of the following situation that
+ * could arise with Mesa-style condition variables: the condition may become
+ * true, making the sleeping thread runnable, but the condition may become
+ * false again before the thread is scheduled. To handle this case, the
+ * condition variable should be used in a while loop as follows:
+ *
```

Comment:
please add markdown style code environment around snippets, here and below.

Reply:
Fixed.